### ISSUE TITLE
Update Drupal version to 9.3

### ISSUE THREAD

**Issue by dnkemper** [uiowa/uiowa#4411]
Blocked until December 8, 2021

Currently unavailable,  [9.3.0 alpha version](https://www.drupal.org/project/drupal/releases/9.3.0-alpha1) is available is not recommended for production sites. Wanted to bring attention to this since a lot of patches I looked at use version 9.3

# Proposed Solution

Wanted to add this as a reminder. Might need Rick to add more since he's the master of this.

- [x] Check dependency compatibility and make note of dependencies that need to be updated at the same time.
- [x] Read all the release notes and make note of any changes that might affect our codebase: https://www.drupal.org/project/drupal/releases/9.3.0 and https://www.drupal.org/project/drupal/releases/9.3.1
- [x] Then test out our functionality
- [x] ~Add/Test: https://www.drupal.org/project/drupal/issues/3195764#:~:text=USER-,Comment,%2317,-dwisnousky%20Credit~ Not needed. Fixed. 
- [x] Remove from composer.json:`"Fix Webform ui button contrast deprecation": "https://www.drupal.org/files/issues/2021-11-07/webform-button-contrast-3195764-12.patch"` - Related to patch above
- [x] Look at other composer.json patches that might be incompatible w/ v9.3

**Comment by pyrello** [comment]
Looks like mid-December is the earliest we would want to do this based on the release schedule: https://www.drupal.org/about/core/policies/core-release-cycles/schedule#current-development-cycle

We'll need to check that our dependencies are compatible before we can go ahead.

**Comment by joewhitsitt** [comment]
Also blocked until 9.3.1 based on https://github.com/uiowa/uiowa/issues/4562#issuecomment-993846550

**Comment by pyrello** [comment]
This is no longer blocked as the issue documented in https://github.com/uiowa/uiowa/issues/4562#issuecomment-993846550 was resolved in 9.3.1. 9.3.2 is now available.

**Comment by pyrello** [comment]
Requires updating the following patches:
- Expose Layout Builder data to REST and JSON:API
- Add Views EntityReference filter to be available for all entity reference fields

**Comment by pyrello** [comment]
Running into an issue where config is being reported as different when running a site sync:
```
 --------------------------- ----------- 
  Name                        State      
 --------------------------- ----------- 
  acquia_connector.settings   Different  
 --------------------------- ----------- 
```

However, running `drush cex` reports no changes. The UI confirms that it is detecting changes, specifically the change in order of config properties that are ignored:
![Screen Shot 2022-02-03 at 1 47 06 PM](https://user-images.githubusercontent.com/709009/152417803-b554230f-67d5-4720-8b51-da958d169c7e.png)

It seems like maybe the `drush cst` command is giving a false positive here? Since the properties are ignored and don't exist in the exported config, it seems their presence in the image above would indicate that they are being added in dynamically. Why they wouldn't be in the correct location when that happens, I have no idea.

**Comment by richardbporter** [comment]
I wonder if this could be removed? https://github.com/uiowa/uiowa/blob/46171986aac9e853a98f32ba387d0b7b006f279c/docroot/modules/custom/uiowa_core/uiowa_core.module#L16-L25

**Comment by richardbporter** [comment]
>// @todo Remove this after https://www.drupal.org/project/drupal/issues/2855675

I should probably read my own comment. 🤦🏻

**Comment by richardbporter** [comment]
> Running into an issue where config is being reported as different when running a site sync:
> 
> ```
>  --------------------------- ----------- 
>   Name                        State      
>  --------------------------- ----------- 
>   acquia_connector.settings   Different  
>  --------------------------- ----------- 
> ```
> 
> However, running `drush cex` reports no changes. The UI confirms that it is detecting changes, specifically the change in order of config properties that are ignored: ![Screen Shot 2022-02-03 at 1 47 06 PM](https://user-images.githubusercontent.com/709009/152417803-b554230f-67d5-4720-8b51-da958d169c7e.png)
> 
> It seems like maybe the `drush cst` command is giving a false positive here? Since the properties are ignored and don't exist in the exported config, it seems their presence in the image above would indicate that they are being added in dynamically. Why they wouldn't be in the correct location when that happens, I have no idea.

Yea that is odd. Running `cim` doesn't seem to do anything either. It's like they are stuck.

**Comment by richardbporter** [comment]
I can get a `sync` to succeed by just ignoring all of `acquia_connector.settings`: 
```
 git diff
diff --git a/config/default/config_ignore.settings.yml b/config/default/config_ignore.settings.yml
index 94ae364a5..e4ba30465 100644
--- a/config/default/config_ignore.settings.yml
+++ b/config/default/config_ignore.settings.yml
@@ -1,9 +1,7 @@
 _core:
   default_config_hash: v7E8C8wJWeAW2BGohMNY1tZSBc4bexM6O62tGxecTfE
 ignored_config_entities:
-  - 'acquia_connector.settings:spi.site_uuid'
-  - 'acquia_connector.settings:spi.site_environment'
-  - 'acquia_connector.settings:spi.site_machine_name'
+  - 'acquia_connector.settings'
   - 'block.block.main_navigation__superfish__horizontal:status'
   - 'block.block.main_navigation__superfish__toggle:status'
   - 'block.block.main_navigation__superfish__mega:status'
```

I think that would be ok to do for now if we wanted. I ran `blt drupal:install` and verified that `acquia_connector.settings` in `config/default` is created for a new site. I did get another config diff after that though:
```
-------------------------- ----------- 
  Name                       State      
 -------------------------- ----------- 
  filter.format.minimal      Different  
  filter.format.plain_text   Different  
 -------------------------- ----------- 
```

**Comment by Ambient-Impact** [comment]
@richardbporter I'm seeing a similar thing on a site I manage that runs 9.3.x and [Configuration Split](https://www.drupal.org/project/config_split), and also happens to be with a filter format config - our staging site continually thinks one of our filter formats has changed, but it's just a different order of keys. That particular config isn't split off but in the main sync directory, so it's been a puzzle.

The only thing that temporarily worked was to do an `sql:sync` from our live/production site, while also exporting config from live and syncing that to staging site at the same time. I say temporarily fixed because it crept back in for some baffling reason. I've just learned to live with it for now, but I would be really curious if someone finds a solution.

**Comment by richardbporter** [comment]
There are a few update hooks and one removes non-existent permissions from roles:
```
>  [notice] The role Authenticated user has had the following non-existent permission(s) removed: access shortcuts.
>  [notice] The role Editor has had the following non-existent permission(s) removed: access file_link_browser entity browser pages, access media_entity_browser entity browser pages, create people block content, create uiowa_articles block content, create uiowa_people block content, delete any people block content, delete any uiowa_articles block content, delete any uiowa_people block content, delete event feed, update any people block content, update any uiowa_articles block content, update any uiowa_people block content.
>  [notice] The role Publisher has had the following non-existent permission(s) removed: access file_link_browser entity browser pages.
>  [notice] The role Webmaster has had the following non-existent permission(s) removed: access file_link_browser entity browser pages, access media_entity_browser entity browser pages, access webform submission log, administer sitenow_find_text configuration, administer uiowa articles, administer uiowa events, assign editor role, assign roles, assign webmaster role, create people block content, create uiowa_articles block content, create uiowa_people block content, delete any people block content, delete any uiowa_articles block content, delete any uiowa_people block content, delete event feed, update any people block content, update any uiowa_articles block content, update any uiowa_people block content.
>  [notice] The roles Authenticated user, Editor, Publisher, Webmaster have had non-existent permissions removed. Check the logs for details.
```

More details in https://www.drupal.org/project/drupal/releases/9.3.0 but we might to double check these.

**Comment by richardbporter** [comment]
> There are a few update hooks and one removes non-existent permissions from roles:
> 
> ```
> >  [notice] The role Authenticated user has had the following non-existent permission(s) removed: access shortcuts.
> >  [notice] The role Editor has had the following non-existent permission(s) removed: access file_link_browser entity browser pages, access media_entity_browser entity browser pages, create people block content, create uiowa_articles block content, create uiowa_people block content, delete any people block content, delete any uiowa_articles block content, delete any uiowa_people block content, delete event feed, update any people block content, update any uiowa_articles block content, update any uiowa_people block content.
> >  [notice] The role Publisher has had the following non-existent permission(s) removed: access file_link_browser entity browser pages.
> >  [notice] The role Webmaster has had the following non-existent permission(s) removed: access file_link_browser entity browser pages, access media_entity_browser entity browser pages, access webform submission log, administer sitenow_find_text configuration, administer uiowa articles, administer uiowa events, assign editor role, assign roles, assign webmaster role, create people block content, create uiowa_articles block content, create uiowa_people block content, delete any people block content, delete any uiowa_articles block content, delete any uiowa_people block content, delete event feed, update any people block content, update any uiowa_articles block content, update any uiowa_people block content.
> >  [notice] The roles Authenticated user, Editor, Publisher, Webmaster have had non-existent permissions removed. Check the logs for details.
> ```
> 
> More details in https://www.drupal.org/project/drupal/releases/9.3.0 but we might to double check these.

Sounds like most of these are probably just not defined in the corresponding module.permissions.yml file: https://www.drupal.org/node/3193348

**Comment by joewhitsitt** [comment]
That is a cool feature and these most of these look like permissions of the past.

Remember Adam's linkit file browser feature or my wild entity inception events/event feed/block stuff?

**Comment by richardbporter** [comment]
Something is deleting this and I'm not sure what: https://github.com/uiowa/uiowa/pull/4741/commits/713246f643addd5b45573ccd9b7196ae03db15f5

Interesting that the media_embed filter has status disable though....

**Comment by richardbporter** [comment]
Might need some help with this change record: https://www.drupal.org/node/3043840

**Comment by bspeare** [comment]
I can take a look at that - is this something that we need to manually check or is there a report that will point out where it is an issue?

**Comment by richardbporter** [comment]
> I can take a look at that - is this something that we need to manually check or is there a report that will point out where it is an issue?

I think its a manual check. The record says:
>This change only affects sites that have enabled configurable display of node base fields (title, uid and created) by means of a hook:

I can't find any instances of that hook in our code so maybe there is nothing to update? I'm not sure, TBH.

**Comment by joewhitsitt** [comment]
@dnkemper do you recall the idea behind these?

> Add/Test: [#3195764: Modules depending on jQuery UI have broken styling since Drupal 9.2](https://www.drupal.org/project/drupal/issues/3195764#:~:text=USER-,Comment,%2317,-dwisnousky%20Credit)

>Remove from composer.json:"Fix Webform ui button contrast deprecation": "https://www.drupal.org/files/issues/2021-11-07/webform-button-contrast-3195764-12.patch" - Related to patch above

Is it just to change to the 9.3 version of the patch and test that it still resolves the button contrast issue? https://www.drupal.org/files/issues/2021-11-06/drupal-suppress_css_from_deprecated_libraries-3195764-16.patch

**Comment by dnkemper** [comment]
@joewhitsitt not at laptop rn but wondering if there was a regression with webforms that was fixed for 9.3... I remember when you add a webform and go to add an element to it, in 9.2 the button contrast was whack. So if we remove the webform contrast patch, hopefully 9.3 fixes that bug?

**Comment by dnkemper** [comment]
@joewhitsitt removed ✅. It fixed the contrast issue. YAY TO UPGRADING!
